### ISSUE TITLE
Collecting of AssertionError messages

### ISSUE THREAD

**Issue by mitzkia** [pytest-dev/pytest#3049]
Hi, 

We are using pytest for our test framework. 
We wrote our "Logger" class to display (on console) and collect (in file) logs from test framework (using logging module). AssertionError displayed on console very convenient, but we also want to collect this information in our log files. 
Our test report dir:
```
- reports/
  - testcase-name/
    - current-date/
       - runtime generated files
       - testcase-log-file.log
```
There was an idea to create assertion methods where we can log with logging module, but we rejected it.
It is better to use only native "assert" instead of someAssertEquals()
So we started to search alternative methods.
For now I checked the logging command line options: log_file, log_level ... and also result-log, but they are not fit completely for our use cases. I noticed that assertions did not captured by logging (it is clear why not), but with a small change it can be done. Can this be a valid case, or something similar?

```
diff --git a/_pytest/runner.py b/_pytest/runner.py
index e07ed2a2..911df0be 100644
--- a/_pytest/runner.py
+++ b/_pytest/runner.py
@@ -4,6 +4,7 @@ from __future__ import absolute_import, division, print_function
 import bdb
 import os
 import sys
+import logging
 from time import time
 
 import py
@@ -302,6 +303,7 @@ def pytest_runtest_makereport(item, call):
             else:  # exception in setup or teardown
                 longrepr = item._repr_failure_py(excinfo,
                                                  style=item.config.option.tbstyle)
+            logging.getLogger('ASSERT').error(longrepr)
     for rwhen, key, content in item._report_sections:
         sections.append(("Captured %s %s" % (key, rwhen), content))
     return TestReport(item.nodeid, item.location,
```
So as I see command line logging options: log_file, log_level ... works with logging module, but did not "capture" AssertionErrors. 
result-log command line option can "capture" AssertionErrors, but not "standard" logging messages.
Please correct me if I am wrong.
As I read there will be a change in logging in 4.0, so maybe it is not too late to think about it :)

For us would be good something mixed functionality of above two way.
We have the following use cases for collecting test logs:
 - we want to display logs on console (with logging module)
 - we also want to collect same logs in files (with logging module)
 - we also want to collect assertion info in files (better if it goes to the same file as normal logs)
 - we want to decide file-path for logs from framework (instead of command line option)
 - we also like colorized logs (eg: colorlog module)

Thanks for the answer.

**Comment by pytestbot** [comment]
[GitMate.io](https://gitmate.io) thinks the contributor most likely able to help you is @nicoddemus.

**Comment by mitzkia** [comment]
As I saw the plugin registering (config.pluginmanager.register(LoggingPlugin(config)) is before the test discovery phase. 
So if I rewrite the "LoggingPlugin" for my own (because of the setup of log_file path), it can not contains the testcase name information, because it is not yet available in this context. 
Is there an other way to inject the testcase name into the log_file information?

Thank you for the answer.

**Comment by mitzkia** [comment]
Sorry, it seems will be work :) 
If I am right the following method will be called before each testcase, so here I can append the testcase name to the path of the --log_file option. (I would replace the original LoggingPlugin with my custom plugin which would contain the testcase name).
```
@pytest.mark.trylast
def pytest_configure(config):
   ...
```

Last question is can the AssertionError message logged within LoggingPlugin? See above code snippet.

Thanks for your answer.

**Comment by mitzkia** [comment]
I would get the tescase name information from: `request.node.name`

**Comment by mitzkia** [comment]
Maybe I can write my own "pytest_runtest_makereport()" in conftest.py, which originally belongs to runner.py.

**Comment by mitzkia** [comment]
Hm..., `pytest_configure()` was not loaded before tests (my fault). 
It was a fixture loading.

**Comment by mitzkia** [comment]
I think I found a solution for log_file naming, somewhere after tests are discovered:
```
my_file_path = "file_path_prefix_%s" % request.node.name.split("/")[-1]
file_handler = logging.FileHandler(my_file_path)
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
file_handler.setFormatter(formatter)

logging_plugin = request.config.pluginmanager.getplugin('logging-plugin')
logging_plugin.log_file_handler = file_handler
```

**Comment by mitzkia** [comment]
Unfortunately, the last solution did not help.
After registering a changed logging-plugin, log messages will go into the initially configured file path (with --log_file=).

**Comment by nicoddemus** [comment]
Hi @mitzkia, sorry for the delay.

IIUC you are trying to capture errors and logging them to a file correct? 

I think you can achieve that by implementing the [pytest_runtest_logreport](https://docs.pytest.org/en/latest/writing_plugins.html#_pytest.hookspec.pytest_runtest_logreport) hook: it will be called  for each test phase (setup/call/teardown) and the report object will contain the traceback if an error occurred. It is just a matter of then logging that information like you want.

**Comment by mitzkia** [comment]
Hi @nicoddemus,

First of all sorry about my previous comments, they can be confused.
Thanks for your answer.
Yes, you are correct.

For better understanding, I would like to summarize my questions.
1, Wanted to log AssertionErrors inside our framework into log files, which are separated by test cases
Found solution (inside teardown):
```
        terminalreporter = self.request.config.pluginmanager.getplugin('terminalreporter')

        if terminalreporter.stats.get('failed'):
            for rep in terminalreporter.stats.get('failed'):
                self.logger.error(str(rep.longrepr))
```
2, Define custom file path for logging-plugin from our framework. It works natively outside with --log-file=[path]
Have tried in one of previous comment: https://github.com/pytest-dev/pytest/issues/3049#issuecomment-353120177
I have not yet found an answer for this question, but for now we can work with our "Logger" class so for now this is not a problem.

Other:
It would be good if:
 - the native logging_plugin module can log/capture also AssertionErrors
 - --log-file path can be defined from own framework, simirally I have tried previously (with help of request.config.pluginmanager.getplugin('logging-plugin') ... ), maybe it already work just I did not find a solution.

Thanks again
